lib/tools/apk-utils.js

```diff
@@ -198,6 +198,47 @@ apkUtilsMethods.install = async function (apk, replace = true, timeout = 60000)
   }
 };
 
+apkUtilsMethods.installOrUpgrade = async function (apk, pkg = null, timeout = 60000) {
+  let apkInfo = null;
+  if (!pkg) {
+    apkInfo = await this.getApkInfo(apk);
+    pkg = apkInfo.name;
+  }
+  if (!pkg) {
+    log.warn(`Cannot read the package name of ${apk}. Assuming correct app version is already installed`);
+    return;
+  }
+  if (!await this.isAppInstalled(pkg)) {
+    await this.adbExec(['install', apk], {timeout});
+    return;
+  }
+  const pkgInfo = this.getPackageInfo(pkg);
+  const pkgVersionCode = pkgInfo.versionCode;
+  if (!apkInfo) {
```

Comment:
this is done intentionally to avoid execution of extra shell command in case the app is not installed.